crates/ibc/src/core/ics04_channel/handler/recv_packet.rs

```diff
@@ -71,13 +71,8 @@ where
         let client_id_on_b = conn_end_on_b.client_id();
         let client_state_of_a_on_b = ctx_b.client_state(client_id_on_b)?;
 
-        // The client must not be frozen.
-        if client_state_of_a_on_b.is_frozen() {
-            return Err(PacketError::FrozenClient {
-                client_id: client_id_on_b.clone(),
-            }
-            .into());
-        }
+        client_state_of_a_on_b.assert_not_frozen()?;
```

Comment:
I'm not sure if I caught you correctly. Wouldn't this only replace a line with `client_state.assert_not_frozen()` with another`get_client_state_validation()`? Can you a bit expand?

Reply:
I'm suggesting to replace

```rust
let client_state_of_a_on_b = ctx_b.client_state(client_id_on_b)?;
client_state_of_a_on_b.assert_not_frozen()?;
```

with 

```rust
// call site
let client_state_of_a_on_b = get_client_state_validation(&ctx_b);


// pseudo-code implementation
fn get_client_state_validation(ctx: &Ctx) {
    let client_state_of_a_on_b = ctx_b.client_state(client_id_on_b)?;
    client_state_of_a_on_b.assert_not_frozen()?;

    client_state_of_a_on_b
}
```

We can probably come up with a better name

Reply:
> let client_state_of_a_on_b = ctx_b.client_state(client_id_on_b)?;
 client_state_of_a_on_b.assert_not_frozen()?;

Isn't the above lines self-explaining enough? If we weren't to use `client_state_of_a_on_b` afterward, having this function would make sense more. but, honestly, I don't value adding a further function here

Reply:
> Isn't the above lines self-explaining enough?

My concern is that after a few refactors we'd forget to check if the client is frozen. Basically, since in validation, we always want to check that the client is not frozen right after receiving it, this new function would be a tool to not forget that.

>  If we weren't to use client_state_of_a_on_b afterward, having this function would make sense more.

What do you mean? `get_client_state_validation` would return the `ClientState` that we just fetch.

Basically I'm looking for a means to ensure that we never forget to do that check over time. Another potential solution would be to ensure that we have a unit test for each handler that will fail if we forget that check.

Reply:
I understand your concern, but I am afraid we achieve this by having this function. We still should be cautious to not miss `get_client_state_validation` wherever needed. Definitely, well-written tests will avoid this issue. 👌🏻

Reply:
Interestingly, with this refactor, we caught test failure in places we missed introducing a `Client` to the `MockContext`.  [425ccb2](https://github.com/cosmos/ibc-rs/pull/531/commits/425ccb2f66280cd0ef8c6234bdc22500711c42a4)

Reply:
> Definitely, well-written tests will avoid this issue. 

I'm good with that. Unless you already verified that we have these tests, we should create an issue to track it.

Reply:
It just captured places we missed the client, but `MockContext` does not cover frozen/expired conditions. 
Opened this https://github.com/cosmos/ibc-rs/issues/538.